### DISCUSSION TITLE
Suspending "agents" from a reactor bank...?

### DISCUSSION THREAD

**Discussion by stustd** [lf-lang/lingua-franca#1842] [Q&A] [answered]
For reactor banks, the input/output cardinality must match. This means (... "is my understanding" ...) that all bank reactors are triggered and must check the presence of their specific (indexed) input. In an application (... which is more of a simulation nature rather than an embedded system ...) I want to use a reactor bank as an agent pool with a bank size equal to the maximum number of simultaneous agents active at any given time. Is there a runtime mechanism that could be used to put agents temporarily to sleep? (I do however anticipate that such an async-like feature may not be compatible with LF's synchronous paradigm ...)

**Comment by edwardalee** [comment]
It is not entirely clear what you are trying to achieve here. First, reactions are only triggered if inputs are present, so one option is to simply not send events to the reactors that you want to be suspended.  This simply moves the logic managing which reactors are suspended to the sender (or an intermediate gateway reactor).  That logic needs to be somewhere.

If instead you are interested in federated execution, where federates are processes that may or may not be running, then @ChadliaJerad is working on a mechanism that supports such "transient federates." It is quite tricky to manage the joining and departing of such transient federates, but she is making great progress.

**Reply by stustd** [reply]
To clarify, it's about a single-node (non-federate) application. It is my understanding that "all reactors in a bank are always triggered, simultaneously, by the vector input message (and therefore all of them consume CPU resources), even if there's only one (indexed) message present meant for only one reactor; i.o.w. all reactors always must check for the presence of "their" (indexed) message and act (or not) accordingly." Although bank reactors can be addressed individually, my question is whether this implies that they're also triggered individually; in my understanding, they're not: they're all triggered, regardless of whether their message is present or not...

(Let me also express my sincere appreciation for the Lingua-Franca technology you and your team are developing.)

**Comment by edwardalee** [comment, marked as answer]
No, this is a misunderstanding of the semantics. A reactor with no present inputs is not triggered, regardless of whether it is in a bank.

**Reply by stustd** [reply]
That is wonderful news; thanks for the clarification!

Also from comparing the MultiPortToMultiPort.lf and MuliBankToMultiBank.lf examples I see that the `is_present` semantics applies to multiports and not to multibanks. The rich examples are really helpful!
